### PR TITLE
refactor: proper modules destruction

### LINKED ISSUE #120: Crash when removing/adding outputs
Example 1:

```
Bar removed from output: DP-3
Bar removed from output: DP-4
Unable to receive IPC response
Unable to receive IPC response

glibmm-CRITICAL **: 12:49:03.096: Error in inter-thread communication: write() failed: Bad file descriptor
Gdk-CRITICAL **: 12:49:03.442: gdk_monitor_get_scale_factor: assertion 'GDK_IS_MONITOR (monitor)' failed
glibmm-CRITICAL **: 12:49:03.443: Error in inter-thread communication: write() failed: Bad file descriptor
Gdk-CRITICAL **: 12:49:03.443: gdk_monitor_get_workarea: assertion 'GDK_IS_MONITOR (monitor)' failed
Gdk-CRITICAL **: 12:49:03.443: gdk_monitor_get_workarea: assertion 'GDK_IS_MONITOR (monitor)' failed
Gdk-CRITICAL **: 12:49:03.443: gdk_monitor_get_scale_factor: assertion 'GDK_IS_MONITOR (monitor)' failed

*lots of "gdk_monitor_get_scale_factor" here*

Tray is in beta, so there may be bugs or even be unusable.

Gdk-CRITICAL **: 12:49:03.543: gdk_monitor_get_scale_factor: assertion 'GDK_IS_MONITOR (monitor)' failed

*lots of "gdk_monitor_get_scale_factor" here*

An object is already exported for the interface org.kde.StatusNotifierWatcher at /StatusNotifierWatcher

Gdk-CRITICAL **: 12:49:03.658: gdk_monitor_get_workarea: assertion 'GDK_IS_MONITOR (monitor)' failed
Gdk-CRITICAL **: 12:49:03.658: gdk_monitor_get_workarea: assertion 'GDK_IS_MONITOR (monitor)' failed

Bar configured (width: 1920, height: 32) for output: eDP-1
zsh: segmentation fault (core dumped)  waybar
```

Example 2:

```
Bar removed from output: DP-6

glibmm-WARNING **: 13:05:46.768: Dropped dispatcher message as the dispatcher no longer exists
glibmm-WARNING **: 13:05:46.768: Dropped dispatcher message as the dispatcher no longer exists

Unable to receive IPC response
corrupted double-linked list
zsh: abort (core dumped)  waybar
```

Honestly, there are a lot of things that can go wrong in my setup:

- Sway itself, which crashes maybe every 3rd - 5th time I mess with outputs.
- Two monitors daisy-chained through DisplayPort, which confuses even Windows sometimes.
- [Kanshi](https://github.com/emersion/kanshi), with a XOR config of either two external monitors or laptop screen.
- Enabled tray module.

So I'm not sure if you'll be able to reproduce this. But yeah, I got Waybar to crash on a second try when turning off and then turning on the second monitor in the daisy chain, seen in example 2. To be clear: I disconnected DP-6, connected it, then disconnected it, and then Waybar crashed when I connected it.

### PR DESCRIPTION
**Pull request by Alexays**
close #120 
@jeremejevs Can you test :)
Also anyone interested if they can test that nothing is broken 😄 
(especially the network that had a big refactor)

### CODE REVIEW

**Comment by jeremejevs** [comment]
@Alexays Thanks for putting time into this! However, sorry, won't be able to test the next few days. You had a repro, right?

**Comment by Alexays** [comment]
@jeremejevs No problem, on my side it works fine, to try it, just unplug a second screen several times and plug it back in, it won't crash anymore but if I forgot something there might be an infinite loop :/

**Comment by Robinhuett** [comment]
Seams to work fine. But my wifi connection gets recognized as ethernet.

File: src/modules/network.cpp [hunk 1]
```diff
@@ -32,73 +25,112 @@ waybar::modules::Network::Network(const std::string& id, const Json::Value& conf
...
+  }
+}
+
+void waybar::modules::Network::createEventSocket()
+{
+  sk_ = nl_socket_alloc();
+	if (genl_connect(sk_) != 0) {
+    throw std::runtime_error("Can't connect to netlink socket");
+  }
+  nl_socket_disable_seq_check(sk_);
```
[the remark this answers is not in the document]
Comment (Robinhuett) [reply src/modules/network.cpp]: No, with `nl_socket_disable_seq_check(sk_);` removed, it still won't populate the essid :/
And no problem, I can test those wifi thingys for you :)

**Comment by Robinhuett** [comment]
@Alexays It's working fine now :)

**Comment by Alexays** [comment]
I'll wait for @jeremejevs feedback and merge this pr and release a new version

**Comment by jeremejevs** [comment]
Hey @Alexays, sorry, never got a chance to test, and I don't have any external monitors anymore, unfortunately. However, I'm happy to pass the gdb baton to @jameswalmsley :smile:

**Comment by Robinhuett** [comment]
Yesterday waybar crashed on me after unplugging my Thinkpad Dock. Unfortunatly I didn't hat time to get a trace. As soon as I cen get to it I'll update here.

**Pull request merged**